src/coreclr/src/System.Private.CoreLib/src/System/Threading/Interlocked.cs

```diff
@@ -2,158 +2,536 @@
+        private static extern int ExchangeAdd(ref int location1, int value);
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern long ExchangeAdd(ref long location1, long value);
+        #endregion
+
+        #region Read
+        /// <summary>Returns a 64-bit signed value, loaded as an atomic operation.</summary>
+        /// <param name="location">The 64-bit value to be loaded.</param>
+        /// <returns>The loaded value.</returns>
+        public static long Read(ref long location) =>
+            CompareExchange(ref location, 0, 0);
+
+        /// <summary>Returns a 64-bit unsigned value, loaded as an atomic operation.</summary>
+        /// <param name="location">The 64-bit value to be loaded.</param>
+        /// <returns>The loaded value.</returns>
+        [CLSCompliant(false)]
+        public static ulong Read(ref ulong location) =>
+            CompareExchange(ref location, 0, 0);
+        #endregion
+
+        #region MemoryBarrier
         /// <summary>
-        /// Implemented: int, long
+        /// Synchronizes memory access as follows:
+        /// The processor that executes the current thread cannot reorder instructions in such a way that memory accesses before
+        /// the call to <see cref="MemoryBarrier"/> execute after memory accesses that follow the call to <see cref="MemoryBarrier"/>.
         /// </summary>
         [Intrinsic]
         [MethodImpl(MethodImplOptions.InternalCall)]
-        internal static extern int ExchangeAdd(ref int location1, int value);
+        public static extern void MemoryBarrier();
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        internal static extern long ExchangeAdd(ref long location1, long value);
+        [DllImport(RuntimeHelpers.QCall, CharSet = CharSet.Unicode)]
+        private static extern void _MemoryBarrierProcessWide();
+
+        /// <summary>Provides a process-wide memory barrier that ensures that reads and writes from any CPU cannot move across the barrier.</summary>
+        public static void MemoryBarrierProcessWide() => _MemoryBarrierProcessWide();
+        #endregion
+
+        #region And
+        /// <summary>Bitwise "ands" two 32-bit signed integers and replaces the first integer with the result, as an atomic operation.</summary>
+        /// <param name="location1">A variable containing the first value to be combined. The result is stored in <paramref name="location1"/>.</param>
+        /// <param name="value">The value to be combined with the integer at <paramref name="location1"/>.</param>
+        /// <returns>The new value stored at <paramref name="location1"/>.</returns>
+        /// <exception cref="NullReferenceException">The address of <paramref name="location1"/> is a null pointer.</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int And(ref int location1, int value)
+        {
+            int current = Volatile.Read(ref location1);
```

Comment:
Does it really need to be `Volatile.Read`? I think regular read would be enough.

Reply:
Just for Int32, or for Int64 as well (and rely on the CE/loop if it gets torn)?

Reply:
`I would do it for both.